### DISCUSSION TITLE
Switching to Mac for iOS Packaging

### DISCUSSION THREAD

**Discussion by FliplineStudios** [airsdk/Adobe-Runtime-Support#3008] [Q&A]
We're planning to release a new game within the next month or two for iOS and Android, and since the App Store has an issue with IPAs packaged from Windows with the latest versions of AIR, I'm trying to wrap my head around what would be involved with switching packaging for iOS over to a Mac for the first time.  Since we've only ever packaged from Windows for our apps, I had a few questions about what's involved -- if someone with more experience on the Mac side of things had any feedback, we'd appreciate it!

1. To use the latest AIR SDK and Apple's linker and so the App Store will accept the build, would we now need to package on a Mac that has and supports Xcode 15 (with iOS 17 SDKs)?

2. Do we need to use Java 11 (as mentioned on airsdk.dev)?  We use this currently for packaging from Windows, wasn't sure if this was still a requirement.

3. Can we continue to use our p12 that we've exported from our iOS Distribution Certificate, or do we need to use the Mac's keychain for signing the IPA on a Mac?

And a couple questions that are mainly for @ajwfrost  :

4. Would we also need to purchase an additional seat for our AIR developer license since we'd be installing the SDK on an additional workstation?  If so, is there a process for adding an additional seat to our account, or do we just start a new purchase order?

5. You had mentioned there wasn't a timeframe yet for getting Apple's ld64 linker adapted again for Windows, but I was curious if you could share any rough idea of whether it's something in the works and aiming for a 50.x update in the (relative) short term, or if due to the complexity it's still unknown or quite a ways away (or aiming for an eventual AIR 51 release) and we should just switch over to a Mac now… Just trying to gauge based on the timing of whether we're better off spending the time to get the new hardware and packaging pipeline set up now for our game release, or waiting a bit longer :)

Our old Mac we use for uploading to the App Store doesn't support the latest Xcode, so if we need to switch our packaging to a Mac, just trying to make sure I know what I'm getting into!

**Comment by ajwfrost** [comment]
Hi @FliplineStudios 

My thinking for these would be:

1. The key thing is to use the Apple linker toolchain so Xcode 14 is needed here. Actually, it might depend whether you need the 'objc_msgSend' support - which is only really needed if you have ANEs that include third party frameworks/libraries such as from Facebook or similar, where they didn't build with the backwards-compatibility flag. So, I would suggest it's worth first using your existing Mac device to see whether it does compile/package up okay with the latest AIR SDK. If you don't get linker errors then you should be okay I think: the actually flags in the IPA package's "info.plist" file, that tell Apple what toolchain and sdk were used, come from the ADT tool so aren't actually related to what you did use ... (although I guess the executable binary would have the details in their load commands somewhere .. it depends how Apple actually check this stuff!)
2. Java 11 was needed on Windows to use the old sun.security classes for code signing. On macOS, if you used a too-recent version of Java, it should then seamlessly switch over to use the built-in `codesign` command available on macOS. So, Java versions aren't such a big thing on mac now..
3. I think using the .p12 is fine for using the ADT code signing (Java 16 or earlier) but if you were using a later version of Java, then the codesign command may want the certificate to be also imported into your login keychain store...
4. Given this is a requirement because of our issues on Windows, I think it would be a bit harsh to require purchasing of another seat for this .. I can make a note here on the system to ensure this doesn't get flagged up a violation (or at least, hopefully to ensure no action is taken when it is flagged - if you do get a message from our support email, just tell them to talk with me!)
5. Timeframes .. are a little uncertain. The earlier Adobe linker build had deliberately excluded the "tbd" functionality, but we're going to need this in order to cope with the way that the same symbol can be in different libraries depending on the different OS versions. I think we're at least a month away from having something that works here. We've just been discussing some potential workarounds, e.g. switching to completely dynamic loading/resolving of symbols, but again we're going to run into the problem with third party binaries that aren't going to have that capability.. we perhaps need to have a bit of a check to see whether there are some simpler options with intermediate libraries that could be created on macOS and then just used on Windows, but I think we may run into the same problems again..

Hope that helps...

**Comment by FliplineStudios** [comment]
@ajwfrost Thanks for your responses!  Sounds like switching to a Mac for packaging is a good idea in the short term for us, we'll move forward with that in case the complicated changes for Windows end up taking a while.  We're due for replacing our old 2011 Mac Mini anyway, so it seems like a good time for newer hardware either way.  Hopefully it'll be a smooth process!

**Comment by FliplineStudios** [comment]
Starting to get packaging up and running on a new Mac, it looks like with 50.2.4.2 we don't have to manually switch things over to ld-classic since it's set up to use that already (instead of LLVM), is that correct?

Is there anything we have to set up for -platformsdk when packaging on a Mac, or will it work without that set?

**Reply by ajwfrost** [reply]
Correct, for 50.2.4.2 on a mac we will automatically look to use ld-classic if we find it, else just the normal ld (with the fallback being the AIR SDK's ld64 binary).

If you use -platformsdk then it will use those libraries as the inputs rather than the SDK-packaged libraries. Which is actually a very good idea, given what Apple have been doing with symbols in different locations. So I would now recommend using that -platformsdk option and setting it to the location of your iPhoneOS SDK within /Applications/Xcode.

thanks

**Comment by FliplineStudios** [comment]
> If you use -platformsdk then it will use those libraries as the inputs rather than the SDK-packaged libraries. Which is actually a very good idea, given what Apple have been doing with symbols in different locations. So I would now recommend using that -platformsdk option and setting it to the location of your iPhoneOS SDK within /Applications/Xcode.

Thanks!  The size of the IPA seems to be virtually the same whether we define -platformsdk or not (only 88 bytes larger when we do), so hopefully everything worked okay with setting that (this fresh install of Xcode had SDK 17.2 so we used that).  The ipa installs fine on iPhones, will try uploading to the App Store soon to make sure nothing gets flagged.  Thanks for the help with this!

**Reply by FliplineStudios** [reply]
Just checking the Info.plist in each of the IPAs, and when I package with -platformsdk pointing to the iPhoneOS17.2.sdk, I still see the same values in Info.plist as when it was packaged using the AIR-included libraries (DTSDKName iphoneos17.0, DTPlatformVersion 17.0), is that to be expected?  Wasn't sure if the platformsdk version was normally reflected in the Info.plist...
